src/Npgsql/NpgsqlConnection.cs

```diff
@@ -751,7 +756,18 @@ public override void EnlistTransaction(Transaction? transaction)
             // Note that even when #1378 is implemented in some way, we should check for mono and go volatile in any case -
             // distributed transactions aren't supported.
 
-            transaction.EnlistVolatile(new VolatileResourceManager(this, transaction), EnlistmentOptions.None);
+            try
+            {
+                var volatileResourceManager = new VolatileResourceManager(this, transaction);
+                transaction.EnlistVolatile(volatileResourceManager, EnlistmentOptions.None);
+                volatileResourceManager.Init();
+            }
+            catch
+            {
+                EnlistedTransaction = null;
```

Comment:
To simplify, why not move `EnlistedTransaction = transaction` to after everything succeeds (after Init), this way you we don't need try/catch? Above, if `transaction == null` we'd need to set it just before returning.